Remove null-timestamp rows that are obsolete after aggregations

daniel-thom opened the pull request:
This PR fixes two problems:
- A derived dataset like comstock_projected has rows with `time_est == null` to designate data that is supposed to be there but is expectedly missing. These are obsolete after dropping the sector and subsector dimensions, but weren't being dropped. Now they are.
- The first fix uncovered an existing problem where weren't properly tracking the column names in the table resulting from a query. This is more of a bandaid than a true fix. We are still treating the column names as "dimension_query_names" even though they are really just labels. One example of why this is still a problem is here: https://github.com/dsgrid/dsgrid/blob/466017cb7dc3bfba667e75019bea65ab3699f658/dsgrid/dataset/table_format_handler_base.py#L80 This function runs after all aggregations and allows users to say they want record names instead of IDs. Using actual dimension_query_names here is convenient. If they added `state` as a column, we can very easily look up the `state` supplemental dimension records. If the user says that they want to use an alias, this will break. We could track this with an additional layer of mappings. Let me know what you think. I'm not sure how much of a problem it is just yet. I could fix it now or defer to later. We could also remove the alias functionality and/or the replace-ids-with-names functionality.

daniel-thom commented on dsgrid/dataset/pivoted_table.py:
```diff
@@ -158,39 +171,62 @@ def process_stacked_aggregations(
-            final_query_names[dim.model.dimension_type].add(column)
-
         for dim_type, columns in final_query_names.items():
+            if dim_type in dropped_dimensions and columns:
```
I did some refactoring while making a fix to this function. The main thing is that when the function exits the columns of the dataframe are stored as "dimension query names" in the query context. But they aren't always actual dimension query names.

elainethale commented:
I don't think we want to remove the alias functionality, because SQL-default names are often not that helpful.

And I don't think we cant to remove the replace ids with names functionality.

If it's not too bad to add the additional layer of mapping needed to support both, I'd like to go ahead and do it now.

daniel-thom commented:
> 

I addressed the issues. The code now stores a list of column_name/dimension_query_name combinations for each column in a query result table. In most cases they will be the same. In cases where the user asks for an alias they will be different. Depending on filtering, the dimension_query_name may not perfectly represent the records in the column. A perhaps more accurate name is "source_dimension_query_name." That is a bit unwieldy and so I stuck with dimension_query_name. If you all want to make a minor change while I'm out, feel free.

daniel-thom commented:
I regenerated the comstock portion of the ReEEDS dataset and it matches the previous version except for the NULL timestamps.

daniel-thom commented:
@elainethale @lixiliu Ready for review.

elainethale commented on dsgrid/query/models.py:
```diff
@@ -146,17 +154,61 @@ class TableFormatType(enum.Enum):
+
+    def replace_metadata(self, dimension_type: DimensionType, metadata: DimensionMetadataModel):
+        """Replace the dimension metadata."""
+        setattr(self, dimension_type.value, metadata)
```
This doesn't have to be 

```
setattr(self, dimension_type.value, [metadata])
```

?

elainethale commented on tests/test_dataset_utils.py:
```diff
@@ -192,6 +196,37 @@ def test_is_noop_mapping_false():
+        )
+        .filter(f"{time_col} IS NOT NULL or count_time == 0")
+        .drop("count_time")
+    )
```
Why is result overwritten with this line that duplicates the logic in the function `remove_invalid_null_timestamps`?

daniel-thom replied:
Oops.

elainethale reviewed (approved):
A couple of minor questions. Thanks for implementing column name tracking!
